minicbor-derive/src/attrs/idx.rs

```diff
@@ -24,12 +24,24 @@ impl Idx {
     }
 
     /// Get the numeric index value.
-    pub fn val(self) -> u32 {
+    fn val(self) -> i32 {
         match self {
             Idx::N(i) => i,
             Idx::B(i) => i
         }
     }
+
+    /// Get the positive index value.
+    pub fn val_as_index(self) -> u32 {
+        self.val()
+            .try_into()
+            .expect("negative index value can only be used as map key, not as array index")
```

Comment:
When collecting the attributes a proper `syn::Error` with a corresponding span should be returned when a negative index is used together with `#[cbor(map)]`. Panicking here is maybe a bit coarse.

Reply:
I only have very little experience with macros: How would I do it better? All call sites can return a syn::Result, but constructing one requires span information. Should that be passed in by the caller, or should the Idx store the span of where it was constructed so that it can return that back out when it fails?

Reply:
I've tried it out with an arbitrary struct that uses default list encoding, putting in a negative number. The error is:

```
error: proc-macro derive panicked
   --> /home/chrysn/git/crates/ariel-os/src/lib/coapcore/src/ace.rs:146:18
    |
146 |         #[derive(minicbor::Encode)]
    |                  ^^^^^^^^^^^^^^^^
    |
    = help: message: negative index value can only be used as map key, not as array index: TryFromIntError(())
```

That's indeed not super pretty, but given that the numbers have to be hardcoded in the definition (no constants supported), it may be good enough.

To make it better, I think I'd need a hint as to how it *should* be done.

Reply:
I am a bit busy at the moment so I can not assist right now. Generally, attributes are collected and checked in `minicbor-derive/src/attrs.rs`.

Reply:
That was a good hint, I found a pattern from there. Now the error message is:

```
error: negative index value can only be used as map key, not as array index
  --> minicbor-tests/tests/encoding.rs:67:9
   |
67 |         #[n(-1)] c: Option<u8>
   |         ^
```

As by-catch I've also fixed the error message for overly large numbers -- not a very realistic case, as it'd only happen on 16-bit hosts, but it was an example I should not have followed.